doc/design/zarr-support-3.md

```diff
@@ -0,0 +1,142 @@
+# Zarr Support
+A zarr archive contains directories and zarr files. 
+Zarr archives are stored in their entirety in S3.
+Zarr archives are represented by a single Asset through the API.
+A tree hash is computed as the zarr archive is uploaded.
+Each node in the tree hash in the hash is stored in S3 to avoid bloating the DB.
+Zarr archives as a whole are too big to be copied.
+
+## Publishing support
+We have not yet finalized a plan for publishing zarr archives, so this design is a stopgap measure to support uploading in the meantime.
+Until the plan is finished and implemented, publishing dandisets with zarr archives will simply not be permitted.
+
+# Requirements
+1. Zarr archives are stored in a "directory" in S3.
+1. Each zarr archive corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr archive.
+1. The API verifies the checksum _immediately_ after upload.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+1. Zarr metadata must be mutable for zarr files in draft dandisets.
+1. Zarr archives must not be copied, as they are too large to reasonably store multiple copies.
+1. Publishing dandisets with zarr archives is not a requirement at this time.
+
+# Implementation
+
+## API endpoints
+* **POST /api/zarr/**
+
+  Create a new zarr archive.
+
+  Returns a zarr ID
+
+* **GET /api/zarr/{zarr_id}/**
+
+  Returns some information about the zarr archive, like name, S3 path+url, and checksum
+
+* **GET /api/zarr/{zarr_id}/upload/**
+
+  Returns a 204 if an upload is in progress, or a 404 if an upload is not in progress.
+
+* **POST /api/zarr/{zarr_id}/upload/**
+
+  Ask to upload a batch of zarr files as part of the zarr archive.
+  No more batch uploads are permitted until this one is completed or aborted.
+
+  Requires a list of file paths and ETags (md5 checksums).
+  The file paths may include already uploaded files; this is how updates are done.
+
+  Returns a list of presigned upload URLs
+
+* **POST /api/zarr/{zarr_id}/upload/complete/**
+
+  Completes an upload of a batch of zarr files.
+  Fail if any of the checksums do not match.
+  Also, recursively update the tree hash for every parent node for each child.
+
+* **DELETE /api/zarr/{zarr_id}/upload/**
+
+  Cancels a batch upload.
+  Any files already uploaded are deleted.
+  A new batch upload can then be started.
+
+* **DELETE /api/zarr/{zarr_id}/files/**
+
+  Deletes zarr files from S3, and updates the tree hash accordingly.
+  Requires a list of file paths
+
+* **POST /api/dandisets/{...}/versions/{...}/assets/zarr/{zarr_id}/**
+
+  Creates a new Asset that points to the zarr archive.
+  Return an asset ID
+
+When added to a dandiset, zarr archives will appear as a normal `Asset` in all the asset API endpoints.
+There will be a flag `zarr: true` indicating that the asset is a zarr archive.
```

Comment:
there should be info in the content type field in the metadata, which is mandatory for an asset. let's just use that instead of adding another flag. also i would be more concerned in general about what happens, if anything, when there is a mismatch, someone uploads zarr and says its a text file. i don't think there is anything that checks.

Reply:
```suggestion
When added to a dandiset, zarr archives will appear as a normal `Asset` in all the asset API endpoints.
```